src/se/vidstige/jadb/server/AdbProtocolHandler.java

```diff
@@ -37,91 +37,111 @@ public void run() {
-                    AdbDeviceResponder device = responder.getDevices().get(0);
-                    output.writeBytes("OKAY");
-                    send(output, device.getType());
-                } else if (command.startsWith("host-serial:")) {
-                    String[] strs = command.split(":",0);
-                    if (strs.length != 3) {
-                        throw new ProtocolException("Invalid command: " + command);
-                    }
-
-                    String serial = strs[1];
-                    boolean found = false;
-                    output.writeBytes("OKAY");
-                    for (AdbDeviceResponder d : responder.getDevices()) {
-                        if (d.getSerial().equals(serial)) {
-                            send(output, d.getType());
-                            found = true;
-                            break;
-                        }
-                    }
-
-                    if (!found) {
-                        send(output, "unknown");
-                    }
-                } else {
-                    throw new ProtocolException("Unknown command: " + command);
-                }
-            } catch (ProtocolException e) {
-                output.writeBytes("FAIL");
-                send(output, e.getMessage());
+        try (
+            DataInputStream input = new DataInputStream(socket.getInputStream());
+            DataOutputStream output = new DataOutputStream(socket.getOutputStream())
+        ) {
+            while (processCommand(input, output)) {
+                // nothing to do here
             }
-            output.flush();
         }
     }
 
+    private boolean processCommand(DataInput input, DataOutputStream output) throws IOException {
+        String command = readCommand(input);
+        responder.onCommand(command);
+
+        try {
+            if ("host:version".equals(command)) {
+                hostVersion(output);
+            } else if ("host:transport-any".equals(command)) {
+                hostTransportAny(output);
+            } else if ("host:devices".equals(command)) {
+                hostDevices(output);
+            } else if (command.startsWith("host:transport:")) {
+                hostTransport(output, command);
+            } else if ("sync:".equals(command)) {
+                sync(output, input);
+            } else if (command.startsWith("shell:")) {
+                shell(input, output, command);
+                return false;
+            } else if ("host:get-state".equals(command)) {
+                hostGetState(output);
+            } else if (command.startsWith("host-serial:")) {
+                hostSerial(output, command);
+            } else {
+                throw new ProtocolException("Unknown command: " + command);
+            }
+        } catch (ProtocolException e) {
+            output.writeBytes("FAIL");
+            send(output, e.getMessage());
+        }
+        output.flush();
```

Comment:
why is this needed?

Reply:
output.flush() was moved from another place (line 121 in current master)

Reply:
👍